Interface for slack's outgoing webhooks and other fixes.

vabs22 opened the pull request.

smarx commented:
*Automated message from Dropbox CLA bot*

@vabs22, it looks like you've already signed the Dropbox CLA. Thanks!

robot-dreams commented on zerver/lib/outgoing_webhook.py:
```diff
@@ -41,23 +41,25 @@ def process_event(self, event):
-    # The returned message will be a dictionary which should have "response_message" as key and response message to
-    # be sent to user as value.
+    # The returned message will be a dictionary which should have atleast
+    # 1. "response_required" as key and boolean True or False as value.
```
Using string keys in this way works, but it is not ideal for a few reasons:
- It's really easy to make a typo (especially if you don't use string constants for the keys)
- It's really easy to forget a key or add an extraneous key
- The type system (MyPy in this case) can't check for safety

Would it be possible to consider some alternative approaches here?  For example:
- Return `None` if no message should be sent
- Return a tuple with 2 elements, where the 2nd element is a boolean that determines whether or not a response should be sent

vabs22 replied:
yeah, your points make sense. I think we can try the tuple method. Though we should think about it carefully at this stage because later changes like this would require changes in all of the interfaces classes. Despite this, I think our tuple approach should work.

robot-dreams replied:
I think the fact that later changes would require changes in all of the interfaces classes is a GOOD thing.  That way, if we update the return value, we can rely on the type system to help us figure out the places we need to update (if it's just a dictionary, it's really easy for us to miss a place).

vabs22 replied:
I agree.

robot-dreams commented on zerver/lib/outgoing_webhook.py:
```diff
@@ -41,23 +41,25 @@ def process_event(self, event):
-    # be sent to user as value.
+    # The returned message will be a dictionary which should have atleast
+    # 1. "response_required" as key and boolean True or False as value.
+    # 2. "response_message" as key and response message to be sent to user as value.
```
I know this is not related to this change, but I'm wondering: what's the motivation for putting the `response_message` into a dictionary, instead of just returning the `response_message` directly as a string?

vabs22 replied:
The main motivation was making it easier for user to return more values. If in future we would require more values than just response_message and response_required, this would be a convenient method.

robot-dreams commented on zerver/models.py:
```diff
@@ -1801,12 +1803,16 @@ class Service(models.Model):
     interface = models.PositiveSmallIntegerField(default=1)  # type: int
 
-    # Valid interfaces are {generic}
+    # Valid interfaces are {generic, zulip_bot_service}
```
Should this also include slack?

vabs22 replied:
yeah, my bad.

robot-dreams commented on zerver/outgoing_webhooks/slack.py:
```diff
@@ -0,0 +1,48 @@
+        if event['message']['type'] == 'private':
+            raise NotImplementedError("Private messaging service not supported.")
+
+        request_data = {"token": event['token'],
```
How did you figure out the list of keys necessary in the `request_data`?  Did you look at some documentation?  If so, can you add the link in a comment?

vabs22 replied:
https://api.slack.com/custom-integrations/outgoing-webhooks, it contains the info about keys.

robot-dreams replied:
Thanks!  I think having this info available as a comment in the codebase will make it much easier next time someone (possibly you) looks at this code.

vabs22 replied:
yeah, I should have added it in the PR description.

vabs22 commented:
@robot-dreams my main concern in this PR is interface for slack. I have took the keys info from slack's documentation and added it here, but I haven't been able to test it for real i.e. sending requests to actual outgoing webhooks and receiving successful response. This is the reason for WIP. 
About UI changes, I have created this [PR](https://github.com/zulip/zulip/pull/5665) which contains the required UI, but has only generic interface. So depending upon which PR gets merged first, I will update the other one appropriately.

vabs22 commented:
@robot-dreams I have updated this PR by changing to tuple instead of dictionary. 
Next if you and @timabbott think the slack's interface is fine, we can go ahead and remove "WIP".

timabbott commented:
@vabs22 I took a look, and the Slack implementation looks reasonable, though I think it's really important that we test that before merging.  What is blocking testing that implementation?

I have two major questions about this implementation:
* I don't understand why you have both the Generic format and ZulipBotServer format; those seem like duplicates?
* I don't understand why you have the `response_required` field?  It seems to me like we should just handle a response if the bot returned one, and have some default return value that doesn't contain a response that bots should use in the event that they don't want to send a response.  That puts the control over whether to respond on the bot code side, so bot authors can customize.

vabs22 commented:
We added 'response_required' primarily to eliminate 2 response messages to the user, one from the bot itself and other from outgoing webhook system. With this, bot would be able to indicate that he would be sending the response itself, so the system doesn't need to give one. Else he can give the system the response to send to the user. Also as a special customization bot would be able to skip any response message.

At the moment, interface for zulip not server is almost same as generic interface. The difference is the response required is no. I am fine with dropping it at the moment and have some other fix to prevent 2 response messages, but I would prefer to keep it this away, to facilitate changes related to zulip bot server easily.

I had tried to find a third party service which services a slack's outgoing webhook, but was unable to, I will try again to try to test it at the earliest. One thing I observed was, slack also sends these values as form post variables as well.

vabs22 commented:
@timabbott @robot-dreams I have tested the slack's interface in real time and it is working. I have now updated the PR and removed WIP. Please take a look.

robot-dreams commented on zerver/lib/outgoing_webhook.py:
```diff
@@ -119,23 +122,32 @@ def do_rest_call(rest_operation, request_data, event, service_handler, timeout=N
+
+def remove_mention_from_message_content(message_content):
+    # type: (Text) -> Text
+    TRIGGER_RE = re.compile(u'(@\*\*.*\*\* )(.*)')
```
- This has the effect of stripping off ALL leading mentions; is this intended?
- If the at mention is in the middle of the message, this will drop everything before the mention; is this intended?  (For example, `hello @**Zulip Bot** how are you?` will become `how are you`)
- Should the trailing space be replaced by `\s+`?
- Are there other mention formats besides `@**mention**` that should be supported?  (The answer might be no)

vabs22 replied:
The main motivation behind this was to make the message more readable to receiver. Some third parties (like where I tested the slack's interface) displayed the message as it is and the message looked a bit confusing to the user. 
I agree with your concerns and I would do more testing to prevent any such errors. 
But I think in the end this parsing functionality would probably be more complex than the above one and it would be much better if the user implemented it itself. But for interfaces like Generic,
I think we can add even simpler parsing logic, like maybe removing just `**` in `@**mention**`. What do you think of this?

vabs22 replied:
Also do you think we should add this parsing logic as a part of interface?

robot-dreams commented on zerver/lib/outgoing_webhook.py:
```diff
@@ -38,28 +39,30 @@ def __init__(self, base_url, token, user_profile, service_name):
-    # be sent to user as value.
+    # The tuple (response_required, response_message) will be returned, where
+    # 1. response_required would be boolean True or False.
+    # 2. response_message would be the response message to be sent to user. It can also be None.
```
First of all, sorry to go back and forth on this.  I'm used to the golang convention where functions often return a tuple like `(string, bool)` or `(string, error)`, but this is because golang doesn't have `None` or exceptions, and @timabbott 's suggestion makes more sense here.

Anyway, if `response_message` can be `None`, and you can use whether or not the message is `None` to determine whether or not a response should be sent, then is it still necessary to include `response_required` in the return value?

vabs22 replied:
Now when I think about it, `response_required` seems unnecessary. Probably I misinterpreted what @timabbott suggested.

robot-dreams commented on zerver/lib/outgoing_webhook.py:
```diff
@@ -119,23 +122,32 @@ def do_rest_call(rest_operation, request_data, event, service_handler, timeout=N
-            succeed_with_message(event, response_data["response_message"])
+            (response_required, response_message) = service_handler.process_success(response, event)
+            if response_required:
+                succeed_with_message(event, response_message)
```
What happens if you call `succeed_with_message` (or `fail_with_message` below) when `response_message` is `None`?

vabs22 replied:
`response_message` is prepended some string, so it will result in error. Are you thinking about moving the logic of checking if it is `None` inside them?

robot-dreams replied:
Yeah, once you remove the `response_required` return value and have `process_event` just return an `Optional[str]` instead of a tuple, I think you can do this instead:

```
if response_message is not None:
  succeed_with_message(event, response_message)
```

vabs22 replied:
Yeah this would be better. Okay, I am on it.

robot-dreams commented on zerver/outgoing_webhooks/generic.py:
```diff
@@ -13,21 +13,19 @@ def __init__(self, base_url, token, user_profile, service_name):
 
     def process_event(self, event):
-        # type: (Dict[str, Any]) -> Tuple[Dict[str ,Any], Dict[str, Any]]
+        # type: (Dict[str, Any]) -> Tuple[Dict[str ,Any], Any]
```
Super nit: Should this be `Tuple[Dict[str, Any], Any]`?

robot-dreams commented on zerver/outgoing_webhooks/generic.py:
```diff
@@ -13,21 +13,19 @@ def __init__(self, base_url, token, user_profile, service_name):
         request_data = {"data": event['command'],
                         "message": event['message']}
-        return (rest_operation, request_data)
+        return rest_operation, json.dumps(request_data)
```
What's the reason for having `json.dumps` here?  Can you please leave a comment?

vabs22 replied:
Initially this was being used inside `do_rest_call`,  during passing the data to `request.request`. It made more sense to me to move this inside the interface. I would add a comment describing this.

timabbott commented:
@vabs22 @robot-dreams what's the status of this effort?  Are there pieces that I should review for merge at this point, or are you two still iterating together?

vabs22 commented on zerver/lib/outgoing_webhook.py:
```diff
@@ -36,32 +37,48 @@ def __init__(self, base_url, token, user_profile, service_name):
+    # Make a message more readable by refactoring @-mentions in a string. EX: it replaces @**vabs22** to
+    # @vabs22 in the message.
+    # The interface classes can also overwrite this function to create their own implementation.
+    def make_message_content_readable(self, message_content):
```
@timabbott I am waiting for views of @robot-dreams about this function. What do you think? Would this be a right approach? The rest seems to be ready.

robot-dreams replied:
Wait, I don't understand why you are doing this.  You originally had code that removed "@-mention"s from the message, and I had some concerns about edge cases in the way you were using regex to accomplish this.

This looks like completely different behavior.  What are you trying to do here?

vabs22 replied:
Here I am trying to replace `@**vabs22**` to `@vabs22`, to make the message more readable. I thought that removing the @-mention altogether could make it confusing to understand, so refactoring it into a more readable form could work. 
Also I have tried some edge cases and it seems to be working. Also it is somewhat better than our previous approach as it replaces and not removes.

robot-dreams replied:
The reason this feels strange to me is because removing `**` seems like a somewhat arbitrary rule, and it's not clear that when you have multiple words, turning `@**Vaibhav Singhal**` into `@Vaibhav Singhal` is what you want.

However, since you have been trying this, I think you should be the one to make this decision, and if you think removing `**` is the best option then I am happy with that one.

However, I would ask that you include a comment about the following:
- Why you do this
- Why it's okay to do this even when the mention contains multiple words
- The fact that this modified message uses a different key (because if this were in `content`, then this could cause problems with the bot library)

robot-dreams commented on zerver/outgoing_webhooks/slack.py:
```diff
@@ -0,0 +1,51 @@
+        # type: (Response, Dict[Text, Any]) -> Optional[str]
+        response_json = json.loads(response.text)
+        response_text = ""
+        if "text" in response_json:
```
Just double checking: in the case where `"text" not in response_json`, do you want to return `""` or `None`?

vabs22 replied:
yeah, I want to send `""` in that case, this would facilitate a basic "Success" response from the system, which is required as there wouldn't be any other response message.

robot-dreams commented on zerver/outgoing_webhooks/slack.py:
```diff
@@ -0,0 +1,51 @@
+        super(SlackOutgoingWebhookService, self).__init__(base_url, token, user_profile, service_name)
+
+    def process_event(self, event):
+        # type: (Dict[str, Any]) -> Tuple[Dict[str ,Any], Any]
```
Super nit: `[str ,Any]` -> `[str, Any]`

vabs22 replied:
okay

timabbott commented:
@vabs22 there's a lot of refactoring in these commits that should really have been split out into independent, smaller changes.  E.g. the change of interface for `get_service_profile` from passing email+realm_id to passing user_profile_id is good, but it should be its own smaller commit.

timabbott commented on zerver/lib/outgoing_webhook.py:
```diff
@@ -81,11 +104,15 @@ def send_response_message(bot_id, message, response_message_content):
 def fail_with_message(event, failure_message):
     # type: (Dict[str, Any], Text) -> None
+    if failure_message is None:
+        return
```
I don't understand why these are checking for `None` and the code is in `process_success` sometimes returns `""`.  This needs to be consistent.  I would use `None` here and modify the `process_success` functions to return `None`, so you're only handling `""` at the boundary interface.

timabbott commented:
This looks generally sane, but doesn't have any tests.  @vabs22 can you add tests that show what format requests will be sent by the Slack format outgoing webhook?

timabbott commented:
Merged via #5896, thanks @vabs22!
